File: app/src/test/java/com/babestudios/companyinfouk/ui/filinghistory/FilingHistoryPresenterTest.java
package com.babestudios.companyinfouk.ui.filinghistory;

import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistoryList;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import rx.Observable;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class FilingHistoryPresenterTest {

	private FilingHistoryPresenter filingHistoryPresenter;

	@Before
	public void setUp() {
		filingHistoryPresenter = new FilingHistoryPresenter(mock(DataManager.class));
		FilingHistoryActivityView view = mock(FilingHistoryActivityView.class);
		when(view.getCompanyNumber()).thenReturn("23");
		when(view.getFilingCategory()).thenReturn("0");
		when(filingHistoryPresenter.dataManager.getFilingHistory("23", "0", "0")).thenReturn(Observable.just(new FilingHistoryList()));
		filingHistoryPresenter.create();
		filingHistoryPresenter.attachView(view);
	}

	@Test
	public void whenGetFilingHistory_thenDataManagerGetFilingHistoryIsCalled() {
		filingHistoryPresenter.getFilingHistory("23", "0");
		verify(filingHistoryPresenter.dataManager, times(2)).getFilingHistory("23", "0", "0");
	}
}

File: app/src/test/java/com/babestudios/companyinfouk/ui/filinghistory/FilingHistoryPresenterTest.kt
package com.babestudios.companyinfouk.ui.filinghistory

import com.babestudios.companyinfouk.CompaniesHouseApplication
import com.babestudios.companyinfouk.DaggerTestApplicationComponent
import com.babestudios.companyinfouk.TestApplicationModule
import com.babestudios.companyinfouk.common.model.filinghistory.CategoryDto
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistory
import io.reactivex.Completable
import io.reactivex.CompletableSource
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class FilingHistoryPresenterTest {

	private lateinit var filingHistoryPresenter: FilingHistoryPresenter

	@Before
	fun setUp() {
		val testApplicationComponent = DaggerTestApplicationComponent.builder()
				.testApplicationModule(TestApplicationModule(CompaniesHouseApplication()))
				.build()
		filingHistoryPresenter = testApplicationComponent.filingHistoryPresenter()
		val viewModel = FilingHistoryViewModel()
		val completable: CompletableSource = Completable.fromCallable { "" }
		filingHistoryPresenter.setViewModel(viewModel, completable)
		`when`(filingHistoryPresenter.companiesRepository.getFilingHistory("23", "", "0")).thenReturn(Single.just(FilingHistory()))
	}

	@Test
	fun whenGetFilingHistory_thenDataManagerGetFilingHistoryIsCalled() {
		filingHistoryPresenter.getFilingHistory("23", CategoryDto.CATEGORY_SHOW_ALL)
		verify(filingHistoryPresenter.companiesRepository, times(1)).getFilingHistory("23", "", "0")
	}

}
